walletsetup.go

```diff
@@ -164,6 +166,18 @@ func createWallet(ctx context.Context, cfg *config) error {
 		// script which provide scripted input to create wallets to
 		// continue working.
 
+		// Ask for a birthday if the wallet was created from seed. If
+		// the wallet is new the birthday is now.
+		if imported {
+			birthday, err = prompt.Birthday(r)
+			if err != nil {
+				return
+			}
+		} else {
+			bd := time.Now()
```

Comment:
This can lead to missed transactions, because the timestamp of the block header can rewind: you set this to `Now()`, generate an address and receive a transaction, but then the block where this is mined gets a timestamp which is before this time. Now on rescans, you won't find this transaction by default.

The timestamp of a block only needs to be greater than the median block timestamp of the last few blocks (see [here](https://github.com/decred/dcrd/blob/63d36ddc2dcb92b59acecea252e5490809b930ff/internal/blockchain/validate.go#L1224-L1229)), which effectively means a block timestamp may be lower than a few of its parents.

This is another reason why I dislike relying on time instead of block for the birthday information... :P

You need to add a time offset here. The median timestamp calculation in consensus code uses 11 blocks (i.e. 55 minutes), but personally, I would just subtract 24h hours from this, just to be on the safe side of any possible issues (a stuck chain or something like that).

Another (better) approach IMO would be to signal that the sync code should select as birthblock the most recent block (again, minus a safety offset) after the first initial sync completes. This would be the most correct way to go about it, because it correctly handles cases where the chain is stuck for extended periods of time.

Reply:
> Another (better) approach IMO would be to signal that the sync code should select as birthblock the most recent block (again, minus a safety offset) after the first initial sync completes. This would be the most correct way to go about it, because it correctly handles cases where the chain is stuck for extended periods of time.

I think this is what I was doing before. So not in `ChainSwitch`?